File: src/dataset/c/triangle_Exercism_chriswilding/gemini/GeminiFlash_triangle_v2.c
#include "triangle.h"

bool is_triangle(triangle_t sides) {
  // Check for non-positive sides early to avoid unnecessary comparisons.
  if (sides.a <= 0 || sides.b <= 0 || sides.c <= 0) {
    return false;
  }
  return sides.a + sides.b > sides.c && sides.b + sides.c > sides.a &&
         sides.c + sides.a > sides.b;
}

bool is_equilateral(triangle_t sides) {
  // Only need to check two sides for equality if it's a valid triangle.
  return is_triangle(sides) && sides.a == sides.b && sides.a == sides.c;
}

bool is_isosceles(triangle_t sides) {
  if (!is_triangle(sides)) {
    return false;
  }
  // Optimized isosceles check: avoids redundant comparisons.
  return sides.a == sides.b || sides.b == sides.c || sides.a == sides.c;
}

bool is_scalene(triangle_t sides) {
  if (!is_triangle(sides)) {
    return false;
  }

  // More efficient scalene check: directly verify that no two sides are equal.
  return sides.a != sides.b && sides.a != sides.c && sides.b != sides.c;
}

File: src/dataset/javascript/triangle_exercism-javascript-oguzsh/gemini/GeminiFlash_triangle_v2.js
export class Triangle {
  constructor(x, y, z) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  isTriangle() {
    return this.x > 0 && this.y > 0 && this.z > 0;
  }

  isEquilateral() {
    if (!this.isTriangle()) {
      return false;
    }
    return this.x === this.y && this.x === this.z;
  }

  triangleInequality() {
    return (
      this.x + this.y > this.z &&
      this.x + this.z > this.y &&
      this.y + this.z > this.x
    );
  }

  isIsosceles() {
    if (!this.isTriangle()) {
      return false;
    }
    if (this.x === this.y || this.x === this.z || this.y === this.z) {
      return this.triangleInequality();
    }
    return false;
  }

  isScalene() {
    return (
      this.triangleInequality() && this.x !== this.y && this.x !== this.z && this.y !== this.z
    );
  }
}